src/main/java/in/projecteka/consentmanager/link/discovery/Discovery.java

```diff
@@ -63,6 +75,63 @@
                                         requestId)));
     }
 
+    public Mono<DiscoveryResponse> patientInHIP(String userName,
+                                              List<PatientIdentifier> unverifiedIdentifiers,
+                                              String providerId,
+                                              UUID transactionId,
+                                              UUID requestId) {
+        return Mono.just(requestId)
+                .filterWhen(this::validateRequest)
+                .switchIfEmpty(Mono.error(ClientError.requestAlreadyExists()))
+                .flatMap(val ->
+                        userWith(userName)
+                                .zipWith(gatewaySystemUrl())
+                                .flatMap(userAndGateway ->
+                                        discoveryServiceClient.requestPatientFor(
+                                                requestFor(userAndGateway.getT1(), transactionId, unverifiedIdentifiers),
+                                                userAndGateway.getT2(),
+                                                providerId))
+                                .zipWith(Mono.delay(Duration.ofSeconds(getExpectedFlowResponseDuration())))
+                                .flatMap(tuple2 ->
+                                        discoveryResults.get(transactionId.toString())
+                                                .switchIfEmpty(Mono.error(ClientError.gatewayTimeOut()))
+                                                .flatMap(dr -> resultFromHIP(dr))
+                ))
+                .switchIfEmpty(Mono.error(ClientError.networkServiceCallFailed()))
+                .flatMap(discoveryResult -> {
+                    if (discoveryResult.getError() != null) {
+                        //Should we get the error and throw client error with the errors
+                        logger.error("[Discovery] Patient care-contexts discovery resulted in error {}", discoveryResult.getError().getMessage());
+                        return Mono.error(ClientError.networkServiceCallFailed());
+                    }
+                    return Mono.just(DiscoveryResponse.builder()
+                            .patient(discoveryResult.getPatient())
+                            .transactionId(transactionId)
+                            .build());
+                })
+                .doOnSuccess(r -> discoveryRepository.insert(providerId, userName, transactionId, requestId));
+
+    }
+
+    private Mono<DiscoveryResult> resultFromHIP(String responseBody) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            return Mono.just(objectMapper.readValue(responseBody, DiscoveryResult.class));
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    private long getExpectedFlowResponseDuration() {
+        return 3L;
```

Comment:
A todo marker to externalize this can help.

Reply:
done